File: advent-of-code-2023/src/main/kotlin/aoc/snowrescuemission/Day01.kt
package aoc.snowrescuemission

import aoc.utils.ChallengeDay
import java.io.File

class Day01(
    fileName: String? = null,
    private val lines: List<String> = fileName?.let { File(it).readLines() } ?: emptyList()
) : ChallengeDay {

    override fun part1() = lines.sumOf { "${it.first(Char::isDigit)}${it.last(Char::isDigit)}".toInt() }
    override fun part2() = lines.sumOf(::part2IntFromFirstAndLastDigit)

    private val digitNames = listOf("one", "two", "three", "four", "five", "six", "seven", "eight", "nine")

    private fun part2IntFromFirstAndLastDigit(line: String): Int {
        val digits = mutableListOf<Int>()
        for ((i, c) in line.withIndex()) {
            if (c.isDigit()) {
                digits.add(c.digitToInt())
            }
            for ((j, name) in digitNames.withIndex()) {
                if (line.substring(i).startsWith(name)) {
                    digits.add(j + 1)
                }
            }
        }
        return "${digits.first()}${digits.last()}".toInt()
    }
}

File: advent-of-code-2024/dart/bin/day01.dart
import 'dart:io';

import 'challenge_day.dart';
import 'utils.dart';

class Day01 implements ChallengeDay {
  final List<String> _lines;

  Day01(this._lines);

  Day01.using(String fileName) : _lines = File(fileName).readAsLinesSync();

  @override
  int part1() {
    return _lines
        .map((e) => toFirstAndLastDigitInt(e))
        .reduce((value, element) => value + element);
  }

  @override
  int part2() {
    var digitNames = [
      "one",
      "two",
      "three",
      "four",
      "five",
      "six",
      "seven",
      "eight",
      "nine"
    ];
    var sum = 0;
    for (var line in _lines) {
      final List<int> digits = [];
      for (int i = 0; i < line.length; i++) {
        var char = line[i];
        if (char.isDigit()) {
          digits.add(int.parse(char));
        }
        for (int j = 0; j < digitNames.length; j++) {
          var name = digitNames[j];
          if (line.substring(i).startsWith(name)) {
            digits.add(j + 1);
          }
        }
      }
      sum += int.parse("${digits.first}${digits.last}");
    }
    return sum;
  }

  int toFirstAndLastDigitInt(String line) {
    var chars = line.split("");
    var first = chars.firstWhere((char) => char.isDigit());
    var last = chars.lastWhere((char) => char.isDigit());
    return int.parse(first + last);
  }

  @override
  String message() => 'day 01: ';
}
